src/components/Nodes/ClientConfig/DynamicConfigForm/FormItem.jsx

```diff
@@ -0,0 +1,147 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+import TextField from '@material-ui/core/TextField'
+import InputAdornment from '@material-ui/core/InputAdornment'
+import IconButton from '@material-ui/core/IconButton'
+import FolderOpenIcon from '@material-ui/icons/FolderOpen'
+import Select from '../../../shared/Select'
+import { Grid as GridAPI } from '../../../../API'
+
+class DynamicConfigFormItem extends Component {
+  static propTypes = {
+    itemKey: PropTypes.string,
+    item: PropTypes.object,
+    client: PropTypes.object,
+    clientName: PropTypes.string,
+    isClientRunning: PropTypes.bool,
+    handleClientConfigChanged: PropTypes.func
+  }
+
+  constructor(props) {
+    super(props)
+    this.inputOpenFileRef = React.createRef()
+  }
+
+  browseDir = key => async event => {
+    const { openFolderDialog } = GridAPI
+    // If we don't have openFolderDialog from Grid,
+    // return true to continue with native file dialog
+    if (!openFolderDialog) {
+      return true
+    }
+    event.preventDefault()
+    // Continue with Grid.openFolderDialog()
+    const { client, clientName } = this.props
+    const defaultPath = client[clientName].config[key]
+    const dir = await openFolderDialog(defaultPath)
+    this.handleChange(key, dir)
+    return null
+  }
+
+  handleChange = (key, value) => {
+    const { handleClientConfigChanged } = this.props
+    handleClientConfigChanged(key, value)
+  }
+
+  render() {
+    const { itemKey, item, client, clientName, isClientRunning } = this.props
+    const label = item.label || itemKey
+    let { type } = item
+    if (!type) type = item.options ? 'select' : 'text'
+    const itemValue = client[clientName].config[itemKey]
+    let options
+
+    switch (type) {
```

Comment:
Nit: each type might be a good candidate for its own `shared` component. Can take place in a future PR.

Reply:
yes agreed, was tempted to break it out but perhaps on adding the next type when it gets too long